proposals/0127-get-sysvar-syscall.md

```diff
@@ -0,0 +1,184 @@
+`SlotHashes::get_slot(slot)` and `StakeHistory::get_entry(epoch)`.
+
+Additionally, as per [SIMD-0088](./0088-enable-core-bpf-programs.md), an effort
+is underway to reduce code complexity of the Solana runtime by porting native
+programs to Core BPF programs. These programs must be 100% backwards compatible,
+therefore requiring access to this sysvar data via syscalls and *not* via
+providing the account directly. Requiring the account would break the programs'
+ABIs.
+
+### Reducing Syscall Bloat
+
+Currently, sysvar data layouts are very difficult to change. When called, the
+syscall handler will copy the entire sysvar to the given memory location.
+
+Notably, BPF programs do not inform the syscall handler of the expected size
+of the sysvar. BPF programs thus effectively assume that the requested sysvars
+will forever have the same size.
+
+This prevents future upgrades that shrink or extend sysvars. In the same
+spirit, if a new sysvar is created, a new syscall must be created.
+
+Furthermore, if certain queries are requred by BPF programs - such as
+`SlotHashes::get_slot()` and `StakeHistory::get_entry()` - those single-element
+retrievals will *also* require new syscalls.
+
+It's worth noting that at the time of this writing, the work required for
+[SIMD-0088](./0088-enable-core-bpf-programs.md) will require at least five new
+syscalls if the existing design is not ratified.
+
+This tightly-coupled relationship between sysvars and syscalls will perpetually
+swell the syscall interface as changes are made, making maintainence and
+reimplementation difficult.
+
+## Alternatives Considered
+
+Other mechanisms considered that provide sysvar data to BPF programs are as
+follows:
+
+### Sysvar-Specific Syscalls
+
+This approach would involve changing nothing. We'd continue on the same course,
+directly coupling new sysvar fetch functions to new syscalls.
+
+### New VM Memory Regions
+
+A new memory region in the VM could be introduced to store sysvar data, which
+could then be made available to BPF programs through an interface around raw
+pointers, which loads the data similarly to the interface proposed herein.
+
+This approach would depend on direct mapping of host memory to VM memory,
+which has been in development for some time and requires further research
+and development to introduce safely to the VM ABI. This extensive R&D period
+would mean Core BPF program initiatives would remain blocked.
+
+Without direct mapping, sysvar data would be fully copied to this memory
+region each time a VM is created (currently one per transaction processed),
+causing a degradation in runtime performance and likely reducing compute
+by far less compared to the compute savings posed by this proposal.
+
+## Detailed Design
+
+By simply redefining the `get` API to accept `offset` and `length` parameters,
+one single `get` syscall can be used by all sysvars to retrieve data.
+
+```rust
+fn get(offset: usize, length: usize) -> &[u8];
+```
```

Comment:
note the current unified `get` api is implementation-defined, built on top of a set of sysvar-specific syscalls. so a new unified syscall would need to accept an argument indicating which sysvar to fetch and read/write from. so we would need a protocol-defined enum mapping sysvars to u8 tags, or we need to pass in the account id of the sysvar and have a protocol-defined list of allowed account ids/sysvars

on that topic im in favor of explicitly not supporting `Instructions` via this mechanism. i think the must-haves are clock, rent, slot hashes, stake history, last restart slot. im less sure about epoch schedule/rewards, assume we dont need slot history, and assume its obvious we dont need fees or recent blockhashes

we should also be clear that sysvars can be added or removed in the future... although in practice i guess we can never really remove them if programs have come to reply on them

also noting that syscalls generally (always?) return a u64 with a success/error code and the output buffer needs to be passed in as a pointer

Reply:
> a new unified syscall would need to accept an argument indicating which sysvar to fetch and read/write from

Good point. I missed that bit. Richie's original draft had the sysvar address being passed. Personally I don't see a reason to opt for the 32-byte address over a single-byte enum, do you? Regardless, I've added this to the proposal.

> on that topic im in favor of explicitly not supporting `Instructions` via this mechanism. i think the must-haves are clock, rent, slot hashes, stake history, last restart slot.

I think it probably makes the most sense to couple this enum, which describes which sysvars are available through the unified syscall, to the sysvar cache. I think if we start to include things from the invoke context in here it might get a bit complicated. In the future, if we decide to apply a unified interface to the invoke context data as well, we can draft a new proposal. What do you think?

I don't see a particular reason to omit anything from the sysvar cache that isn't deprecated - like the epoch schedule/rewards. If they're there, might as well make them available.

> we should also be clear that sysvars can be added or removed in the future... although in practice i guess we can never really remove them if programs have come to reply on them

I touched on adding new ones a few times in here, but yeah I specifically didn't say remove because that would be a deprecation process for the forward-facing sysvar interface, not this unified syscall.

> also noting that syscalls generally (always?) return a u64 with a success/error code and the output buffer needs to be passed in as a pointer

True, I was trying to psuedo-code demonstrate that you'd provide those params and get a slice, but you're right I should probably be a bit more explicit on how the syscall-sysvar relationship would work.

Reply:
Pushed changes with this feedback addressed!